### PR TITLE
Move references in docstrings

### LINKED ISSUE #191: Use one reference file for docstrings
As stated in #190, use 1 reference file for all references in the docstrings. 

Helps avoid the additional step of adding references for each function as well as duplicated reference errors. 

https://mystmd.org/guide/citations#including-bibtex

https://sphinxcontrib-bibtex.readthedocs.io/en/latest/usage.html

### PR DESCRIPTION
**Pull request by purva-thakre**
## Description

Fixes #191 

## Todos
Notable points that this PR has either accomplished or will accomplish.
  -  [ ] TODO 1

## Questions
-  [ ] Question1

## Status
-  [ ] Ready to go

### CODE REVIEW

File: docs/_autosummary/toqito.state_props.is_separable.rst [hunk 1]
```diff
@@ -1,29 +1,6 @@
...
-   
-
-   
-   
-   
-
-
+.. currentmodule:: toqito.state_props
 
+.. autofunction:: is_separable
```
Comment (purva-thakre) [inline docs/_autosummary/toqito.state_props.is_separable.rst:6]: Probably want to undo this change.

File: docs/_autosummary/toqito.channel_metrics.completely_bounded_spectral_norm.rst [hunk 2]
```diff
@@ -0,0 +1,6 @@
+﻿toqito.channel\_metrics.completely\_bounded\_spectral\_norm
+===========================================================
+
+.. currentmodule:: toqito.channel_metrics
+
+.. autofunction:: completely_bounded_spectral_norm
```
Comment (purva-thakre) [inline docs/_autosummary/toqito.channel_metrics.completely_bounded_spectral_norm.rst:6]: @vprusso `autodocsummary` is adding these extra files because they don't exist. For this file, there should have been something at `docs/_autosummary/toqito.channel_metrics.completely_bounded_spectral_norm.rst`. 

Should I keep these or undo this commit?

https://github.com/vprusso/toqito/blob/master/docs/_autosummary/toqito.channel_metrics.completely_bounded_spectral_norm.rst

File: docs/conf.py [hunk 3]
```diff
@@ -41,8 +41,12 @@
     "sphinx.ext.viewcode",
     "sphinx.ext.githubpages",
     "sphinx.ext.napoleon",
+    "sphinxcontrib.bibtex",
 ]
 
+bibtex_bibfiles = ["books.bib", "articles.bib"]
+bibtex_default_style = "unsrt"
+suppress_warnings = ["bibtex.duplicate_label", "bibtex.duplicate_citation", "bibtex.duplicate_id"]
```
Comment (purva-thakre) [inline docs/conf.py:49]: Comment with reasons for why we disable this.

**Comment by purva-thakre** [comment]
@vprusso Are you fine with how QETLAB is cited throughout the references? Thought I should double-check before I do the same for others. 

image

**Comment by vprusso** [comment]
> @vprusso Are you fine with how QETLAB is cited throughout the references? Thought I should double-check before I do the same for others. 
> 
> 
> 
> image
> 
> 

Yep, I think that seems fine to me!

**Comment by purva-thakre** [comment]
Note `make clean html` currently fails due to following error. Not sure where the braces are unbalanced because the error message does not pinpoint to the exact spot.

```
Extension error (sphinxcontrib.bibtex.domain):
Handler <function env_updated at 0x7f16d9045580> for event 'env-updated' threw an exception (exception: syntax error in line 1: unbalanced braces)
make: *** [Makefile:20: html] Error 2
```

**Comment by vprusso** [comment]
> Note `make clean html` currently fails due to following error. Not sure where the braces are unbalanced because the error message does not pinpoint to the exact spot.
> 
> ```
> Extension error (sphinxcontrib.bibtex.domain):
> Handler <function env_updated at 0x7f16d9045580> for event 'env-updated' threw an exception (exception: syntax error in line 1: unbalanced braces)
> make: *** [Makefile:20: html] Error 2
> ```

Hmm, that error isn't very helpful. 

I attempted running:

```
make clean && make html SPHINXOPTS="-nWT --keep-going -vvv"
```

in the `docs/` folder to see a bit more info on the suggestion of this [post](https://github.com/numpy/numpydoc/issues/268). They seemed to be encountering the same error. 

It's possible it could be some malformed docstring, but I wonder if it's something broken on the Sphinx side of things. 

One thing that might be worthwhile to check is to create a separate branch that just alters the bib information of a single file. If that still breaks, it's likely there is something still broken on the Sphinx side of things.

**Comment by purva-thakre** [comment]
> It's possible it could be some malformed docstring, but I wonder if it's something broken on the Sphinx side of things.

I did keyword search for `{` and `}` to make sure the braces were balanced properly in the bib files. Number of left braces is exactly the same as the number of right braces.

I will give it a try by trying out the changes in a new branch.

**Comment by purva-thakre** [comment]
Actually, no. I take back what I said in my previous comment. 

`pybtex` is flagging an unbalanced braces issue. The bad error message is by `pybtex`. 

https://bitbucket.org/pybtex-devs/pybtex/src/2b82780b5090b1c3de0875617b782fbb5b74ba4e/pybtex/markup/__init__.py#lines-62

**Comment by vprusso** [comment]
> Actually, no. I take back what I said in my previous comment.
> 
> `pybtex` is flagging an unbalanced braces issue. The bad error message is by `pybtex`.
> 
> https://bitbucket.org/pybtex-devs/pybtex/src/2b82780b5090b1c3de0875617b782fbb5b74ba4e/pybtex/markup/__init__.py#lines-62

Gotcha. Good to know, but I'm wondering if this helps us narrow it down any further. 

As an experiment, is it straightforward to do the following:

- Create a new branch from `master`
- Add in a minimal set of changes (just the alterations you made in `pyproject.toml` and the files in `docs/` (excluding the ones in `_autosummary`
- Check if `make clean && make html` works

If it does, this might indicate that somewhere amongst the changes there is some badly formatted doc. If it doesn't, this might indicate that either there is something broken within one of the sphinx dependencies and either needs to be updated on our end (by invoking the right versions) or on their end (which is in some sense out of our control). 

Hopefully, it illicits the former which would make it easier for us to hone in on what/where the ill-formatted docstring lives. 

Does that seem reasonable? Happy to consider other alternatives as well if you have other thoughts/ideas.

**Comment by purva-thakre** [comment]
Reopened this PR. The issue was with pybtex not being able to parse some URLs as URLs. For now, all the URLs that caused issues are strings. Will create a new issue for this.

**Comment by purva-thakre** [comment]
@vprusso This is ready for a review after resolving the conflicts.

File: toqito/channel_metrics/diamond_norm.py [hunk 4]
```diff
@@ -7,13 +7,15 @@ def diamond_norm(choi_1: np.ndarray, choi_2: np.ndarray) -> float:
     r"""
     Return the diamond norm distance between two quantum channels.
 
-    The calculation uses the simplified semidefinite program of Watrous in [CBN]_.
+    The calculation uses the simplified semidefinite program of Watrous in
+    :cite:`Watrous_2009_semidefinite`.
 
-    This function has been adapted from: https://github.com/rigetti/forest-benchmarking
+    This function has been adapted from :cite:`Rigetti_docs`
```
Comment (vprusso) [inline toqito/channel_metrics/diamond_norm.py:13]: Maybe `Rigetti_docs` should be `rigetti2022forest` or something like that?

File: toqito/channel_ops/choi_to_kraus.py [hunk 5]
```diff
@@ -12,7 +12,7 @@ def choi_to_kraus(
     choi_mat: np.ndarray, tol: float = 1e-9, dim: int | list[int] | np.ndarray = None
 ) -> list[np.ndarray] | list[list[np.ndarray]]:
     r"""
-    Compute a list of Kraus operators from the Choi matrix [Rigetti20]_.
+    Compute a list of Kraus operators from the Choi matrix from `Rigetti <https://github.com/rigetti/forest-benchmarking>`__.
```
Comment (vprusso) [inline toqito/channel_ops/choi_to_kraus.py:15]: We have a link to the forest benchmarking docs here but elsewhere we have a reference to the bibtex key for forest. Either is fine, but should be consistent.

File: toqito/channel_ops/choi_to_kraus.py [hunk 6]
```diff
@@ -24,7 +24,8 @@ def choi_to_kraus(
     For completely positive maps the output is a single flat list of numpy arrays since the left and
     right Kraus maps are the same.
 
-    This function has been adapted from [Rigetti20]_ and QETLAB package.
+    This function has been adapted from `Rigetti <https://github.com/rigetti/forest-benchmarking>`__ 
+    and QETLAB :cite:`QETLAB_Kraus_Operators`.
```
Comment (vprusso) [inline toqito/channel_ops/choi_to_kraus.py:28]: Maybe we just have one reference for QETLAB instead of multiple ones for each function/feature?

File: toqito/channel_props/is_unital.py [hunk 7]
```diff
@@ -15,7 +15,7 @@ def is_unital(
     dim: int | list[int] | np.ndarray = None,
 ) -> bool:
     r"""
-    Determine whether the given channel is unital [WatUnital18]_.
+    Determine whether the given channel is unital (Chapter: Unital channels and Majorization from :cite:`watrous_2018`).
```
Comment (vprusso) [inline toqito/channel_props/is_unital.py:18]: "channels" should be capitalized.

File: toqito/channels/reduction.py [hunk 8]
```diff
@@ -37,10 +37,16 @@ def reduction(dim: int, k: int = 1) -> np.ndarray:
...
+        :filter: docname in docnames
+
     :param dim: A positive integer (the dimension of the reduction map).
     :param k: If this positive integer is provided, the script will instead return the Choi
               matrix of the following linear map: Phi(X) := K * Tr(X)I - X.
     :return: The reduction map.
     """
     psi = max_entangled(dim, False, False)
-    return k * identity(dim**2) - psi * psi.conj().T
+    return k * identity(dim ** 2) - psi * psi.conj().T
```
Comment (vprusso) [inline toqito/channels/reduction.py:52]: Actually, if you've already made a change to this line, I believe the more proper way to do multiplication is by way of the `@` symbol. When I started `toqito`, this wasn't the case, but I've been trying to replace `*` with `@`where appropriate (i.e. anywhere there is vector-to-vector or vector-to-matrix multiplication).

```suggestion
return k * identity(dim ** 2) - psi @ psi.conj().T
```

File: toqito/helper/np_array_as_expr.py [hunk 9]
```diff
@@ -6,8 +6,7 @@
 
 
 def np_array_as_expr(np_arr: [np.ndarray]) -> Expression:
```
Comment (vprusso) [inline toqito/helper/np_array_as_expr.py:8]: Hmm. What's our comment docstring convention? Should we begin the comment at the end of the `"""` or start a new line?

I think either is okay, but we should be consistent. The pedantic part of me wants to go through and do the former everywhere, but that's another issue for another time :)

File: toqito/helper/npa_hierarchy.py [hunk 10]
```diff
@@ -128,7 +128,7 @@ def npa_constraints(  # pylint: disable=too-many-locals
     assemblage: dict[tuple[int, int], cvxpy.Variable], k: int | str = 1, referee_dim: int = 1
 ) -> list[cvxpy.constraints.constraint.Constraint]:
     r"""
-    Generate the constraints specified by the NPA hierarchy up to a finite level.
+    Generate the constraints specified by the NPA hierarchy up to a finite level :cite:`Navascu_s_2008`.
```
Comment (vprusso) [inline toqito/helper/npa_hierarchy.py:131]: Hmm, that's a weird bibtex key. Maybe if we follow the prior comment and uniformize these things, the weirdness here will go away though, so this comment might just be moot.

File: toqito/matrices/gen_pauli.py [hunk 11]
```diff
@@ -16,7 +16,7 @@ def gen_pauli(k_1: int, k_2: int, dim: int) -> np.ndarray:
     from 0 to :code:`dim-1`, inclusive.
 
     Note that the generalized Pauli operators are also known by the name of
-    "discrete Weyl operators". [WatrousLec6]_
+    "discrete Weyl operators". (Lecture 6: Further remarks on measurements and channels from :cite:`Watrous_2011_lecture_notes`)
```
Comment (vprusso) [inline toqito/matrices/gen_pauli.py:19]: Should we capitalize the title of this lecture or leave it as is?

File: toqito/matrix_ops/vectors_to_gram_matrix.py [hunk 12]
```diff
@@ -3,7 +3,11 @@
 
 
 def vectors_to_gram_matrix(vectors: list[np.ndarray]) -> np.ndarray:
-    """Given a list of vectors, return the Gram matrix."""
```
Comment (vprusso) [inline toqito/matrix_ops/vectors_to_gram_matrix.py:6]: I think one of my diffs takes care of the more in-depth docstring: 
https://github.com/vprusso/toqito/pull/265

File: toqito/state_props/tests/test_entanglement_of_formation.py [hunk 13]
```diff
@@ -9,24 +9,30 @@
...
-    # Invalid non-square matrix for entanglement_of_formation.
-    (np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9], [10, 11, 12]]), None),
-])
+@pytest.mark.parametrize(
+    "rho, dim",
+    [
+        # Invalid local dimension for entanglement_of_formation.
+        (np.identity(4), 3),
+        # Not presently known how to calculate for mixed states.
+        (3 / 4 * e_0 * e_0.conj().T + 1 / 4 * e_1 * e_1.conj().T, None),
```
Comment (vprusso) [inline toqito/state_props/tests/test_entanglement_of_formation.py:31]:
```suggestion
(3 / 4 * e_0 @ e_0.conj().T + 1 / 4 * e_1 @ e_1.conj().T, None),
```

File: toqito/states/isotropic.py [hunk 14]
```diff
@@ -50,13 +50,13 @@ def isotropic(dim: int, alpha: float) -> np.ndarray:
...
+        :filter: docname in docnames
+
 
     :param dim: The local dimension.
     :param alpha: The parameter of the isotropic state.
     :return: Isotropic state of dimension :code:`dim`.
     """
     psi = max_entangled(dim, False, False)
-    return (1 - alpha) * np.identity(dim**2) / dim**2 + alpha * psi * psi.conj().T / dim
+    return (1 - alpha) * np.identity(dim ** 2) / dim ** 2 + alpha * psi * psi.conj().T / dim
```
Comment (vprusso) [inline toqito/states/isotropic.py:62]:
```suggestion
    return (1 - alpha) * np.identity(dim ** 2) / dim ** 2 + alpha * psi @ psi.conj().T / dim
```

**Review by vprusso — APPROVED** [review]
Minor comments, but overall, LGTM modulo those comments!

File: toqito/channel_metrics/completely_bounded_spectral_norm.py [hunk 15]
```diff
@@ -6,20 +6,14 @@
 
 
 def completely_bounded_spectral_norm(phi: np.ndarray) -> float:
-    r"""
-    Compute the completely bounded spectral norm of a quantum channel.
+    r"""Compute the completely bounded spectral norm of a quantum channel.
```
[the remark this answers is not in the document]
Comment (purva-thakre) [reply toqito/channel_metrics/completely_bounded_spectral_norm.py:9]: Line length > 120 if I have the citations on the same line.

File: toqito/channel_ops/choi_to_kraus.py (hunk 6 above)
Comment (purva-thakre) [reply toqito/channel_ops/choi_to_kraus.py:28]: Fine by me.

File: toqito/channel_metrics/completely_bounded_spectral_norm.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (purva-thakre) [reply toqito/channel_metrics/completely_bounded_spectral_norm.py:9]: line length > 120 if I do this

File: toqito/channel_metrics/completely_bounded_trace_norm.py [hunk 16]
```diff
@@ -12,23 +12,15 @@ def completely_bounded_trace_norm(phi: np.ndarray) -> float:
     Find the completely bounded trace norm of a quantum channel.
 
     Also known as the completely bounded diamond norm of a quantum
-    channel [WatCNorm18]_. The algorithm in p.11 of [WatSDP09] with
-    implementation in QETLAB [JohQET] is used.
+    channel (Section 3.3.2 of :cite:`watrous_2018`). The algorithm in p.11 of :cite:`watrous2012simpler` with
```
[the remark this answers is not in the document]
Comment (purva-thakre) [reply toqito/channel_metrics/completely_bounded_trace_norm.py:15]: yes, I created a new issue for this #276 . I was lazy and mostly used what was provided in the bib info by arxiv because the changes are becoming too big to keep track of in this PR.

File: toqito/helper/npa_hierarchy.py (hunk 10 above)
Comment (purva-thakre) [reply toqito/helper/npa_hierarchy.py:131]: Yep, this was the key from arxiv's bib info. I was lazy here as well.

File: toqito/helper/np_array_as_expr.py (hunk 9 above)
Comment (purva-thakre) [reply toqito/helper/np_array_as_expr.py:8]: How about we leave this for #280? These types of inconsistencies are currently being caught locally.

File: docs/articles.bib [hunk 17]
```diff
@@ -1,15 +1,16 @@
 #File for all article, arxiv and other web references
+# To add a key, Lastname_year_abbreviatedtitle/firstletteroftitle
```
Comment (purva-thakre) [inline docs/articles.bib:2]: > To add a key, Lastname_year_abbreviatedtitle/firstletteroftitle

@vprusso Would this work for the sake of consistency?

File: toqito/channel_metrics/completely_bounded_spectral_norm.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (vprusso) [reply toqito/channel_metrics/completely_bounded_spectral_norm.py:9]: Ah, gotcha. Makes sense!

File: docs/articles.bib (hunk 17 above)
Comment (vprusso) [reply docs/articles.bib:2]: We could align with how [Google Scholar does bibtex keys](https://scholar.googleusercontent.com/scholar.bib?q=info:wK_Z0wl5wBoJ:scholar.google.com/&output=citation&scisdr=ClE5Oj9PEMCi-fd4zM0:AFWwaeYAAAAAZWF-1M36GYDDR2lzgkE_gMNj-jo&scisig=AFWwaeYAAAAAZWF-1Ft1DjFeYXIc9LVPOBdjjyU&scisf=4&ct=citation&cd=-1&hl=en), which is mostly what you have there. That is `{lastname}{year}{abbreviatedtitle}` Any opposition to that?

Comment (purva-thakre) [reply docs/articles.bib:2]: I think I am going to create an issue for using Google scholar style for bibtex keys. I'll get to making these changes then.
